### ISSUE TITLE
Memory leak using findAll method

### ISSUE THREAD

**Issue by evenfrost** [sequelize/sequelize#12948]
### What are you doing?

Iterating over Postgres DB data via `findAll` method with `limit` and `offset` params. Here is the minimal reproducible code:

```ts
import { Model } from './models';

void (async () => {
  let currentOffset = 0;

  const queryData = async (offset = 0) => {
    await Model.findAll({
      limit: 10000,
      offset,
    });

    await queryData(currentOffset += 10000);
  };

  await queryData();
})();

```

### What do you expect to happen?

The code runs fine and doesn't cause memory leaks.

### What is actually happening?

Node.js process grows in memory consumption over time and causes a memory leak and process exit.

```
Output here
```

### Additional context

Here is the [similar issue](https://github.com/sequelize/sequelize/issues/8145), which was closed by the bot.

### Environment

- Sequelize version: 6.3.5
- Node.js version: 14.15.4
- Operating System: Ubuntu 20.04
- TypeScript version: 4.1.3

### How does this problem relate to dialects?

- [x] I think this problem happens regardless of the dialect.
- [ ] I think this problem happens only for the following dialect(s): 
- [ ] I don't know, I was using PUT-YOUR-DIALECT-HERE, with connector library version XXX and database version XXX

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [x] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by SpeedyCraftah** [comment]
I think you have just completely discovered my memory leak issue I've been trying to debug for months! We use Sequelize and have a lot of `findAll` queries so this totally caught my eye. I'll try to see if I can reproduce the issue and see if that was in fact the problem the whole time and possibly dive into the source code for a fix.

**Comment by SpeedyCraftah** [comment]
I think I can confirm.

![image](https://user-images.githubusercontent.com/45142584/105069488-eab1de00-5a79-11eb-8d5e-52bcfdb71aa1.png)
![image](https://user-images.githubusercontent.com/45142584/105069492-ebe30b00-5a79-11eb-84e8-5b2a1462a9d5.png)

**Comment by nawlbergs** [comment]
Just from looking at that... every function is waiting for its child call to resolve.. which never happens.. because every child promise returns another promise... infinitely...  it might not be the call itself.. but the fact your infinitely nesting a promise... the parents never resolve and cant be garbage collected?

**Comment by SpeedyCraftah** [comment]
> Just from looking at that... every function is waiting for its child call to resolve.. which never happens.. because every child promise returns another promise... infinitely... it might not be the call itself.. but the fact your infinitely nesting a promise... the parents never resolve and cant be garbage collected?

You have a point and I would like to see @evenfrost try the same code without awaiting the `queryData` function call at the bottom of the function. Even though this may be a cause it might not fix the problem completely since I'm experiencing the same issue by just calling `findAll` in a manner where it gets disposed when it's no longer needed.

**Comment by evenfrost** [comment]
Actually, I rewrote it with `while` loop (e.g. `while (hasMoreData) { do await stuff }`, and now it seems to work fine. I believe @nawlbergs suggestion may be true but still want to know what the best approach here is.

**Comment by alexnault** [comment]
I ran some load testing.

Here's a heap snapshot before, with Sequelize taking around 1.5MB

<img width="1398" alt="select1" src="https://user-images.githubusercontent.com/6646230/118275891-014ecb80-b495-11eb-8e95-0d0d1c984099.png">

And then here's **6 hours after the load testing was over**:

<img width="1409" alt="select" src="https://user-images.githubusercontent.com/6646230/118275384-5a6a2f80-b494-11eb-9b79-aebe62ac5c5e.png">

As you can see, Sequelize hasn't released 114MB worth of memory.

[handleSelectQuery()](https://github.com/sequelize/sequelize/blob/9f950cbcbdd659d559496b77c40e0f827b108561/lib/dialects/abstract/query.js#L244) seems to be the culprit.

Note: I am also using Postgres

**Comment by fh-cmays** [comment]
I am experiencing this issue as well.

**Comment by sam-lord** [comment]
Me too, using Sequelize and MySQL

**Comment by sachinda123456** [comment]
Is there any work around for this issue?
Also would it be safe to use this package within a production workload?

**Comment by sachinda123456** [comment]
any idea

**Comment by SpeedyCraftah** [comment]
why does this system even exist

**Comment by WikiRik** [comment]
The bot should not mark this as stale every 14 days anymore

**Comment by hironate** [comment]
We are facing production issue, because of this memory leak, our containers are getting restarted in load environment when it reaches to peak memory. Is anybody going to fix this issue or not ?

**Comment by ephys** [comment]
I haven't been able to replicate the problem. The issue in the OP isn't caused by Sequelize but by the infinitely recursing function.

Without further data or an examples that does have a memory leak, we can't fix this

**Comment by hironate** [comment]
Let me send you boilerplate repo with docker and details how you can reproduce them.

**Comment by hironate** [comment]
@ephys here is a repo and way to reproduce https://github.com/hironate/sequelize-memory-leak

**Comment by ephys** [comment]
I couldn't manage to reproduce a memory leak. I queries `/get` 3443 times, each time it returned 57 items. The memory usage fluctuated between 19MB and 35MB while the queries were being run then once I stopped it fell back to 19MB

![image](https://user-images.githubusercontent.com/1280915/194900970-6f9f38eb-cef5-4e25-b052-e789aac3e372.png)

Exploring the heap, the only new elements in the heap related to sequelize after running the 3443 queries were `Executing (default): SELECT "id", "userName", "email", "password", "createdAt", "updatedAt" FROM "Users" AS "User";` which was still in memory because it was used by chrome devtools. Clearing the devtools purged all of those strings from memory

**Comment by hironate** [comment]
Thanks @ephys let me check things from my side.

**Comment by davidrapin** [comment]
We have a case of flaky memory leak, related to sequelize calls.
In our case, I suspect that the issue is with `findAll` calls that involve eagerly loaded associations (my best guess is circular references between linked objects).

**Comment by cpan1995** [comment]
> We have a case of flaky memory leak, related to sequelize calls. In our case, I suspect that the issue is with `findAll` calls that involve eagerly loaded associations (my best guess is circular references between linked objects).

we had a similar problem with associations. The current work around for us is to break out the association and call the each id inside a loop. Not fast, not ideal but the only thing that seems to work for us.

**Comment by BBC6BAE9** [comment]
experiencing this issue with findAll.

**Comment by Hyjaz** [comment]
I've had a similar problem when using findOne with through more than a million rows. The query looked like the following:

```
   const condition: WhereOptions<ActiveUser> = {
      user_id: userId,
    };
    
    ActiveUser.findOne({
      where: condition,
    });
  }
  ```
  
  This caused memory leaks up to 5gb and then OOMed.
  
  I converted the above to a raw sql query `Select ... limit 1` and it is now using 47mb.
  
  There definitely is something leaking in the findAll. I am not using association either. The simplest of tables but just with a lot of rows.
  
  I will be modifying all my findAlls and findOnes to raw sql queries.

**Comment by lucidity-mark** [comment]
I was experiencing the same memory leak issues with .findAll mentioned above and updating from Node 14 to 18 fixed it for me.

**Comment by ephys** [comment]
Unfortunately without a reproducible example there is no way for us to replicate your issue. We've tried multiple times in the past and are yet to find any evidence of a memory leak :/

**Comment by Hyjaz** [comment]
So if you do a findOne() over and over again on a primary key that has millions of rows it doesn't run into memory leak problems?

**Comment by ephys** [comment]
Not the last time I ran it, the memory usage remained constant

**Comment by hesar** [comment]
I am facing right now same issue for findAll()
this is completely fresh install
node: 18.14
sequelize: 6.14.0
tedious: 14.1.0
mssq driver: 9.1.1


I have MSSQL  DB on SQL Express having over 1 000 000 rows and around 40 colums
Execution time in  SQL server mgmt studio takes around 14 seconds

simple code 
```
routes.get('/DB2023', (req, res) => {
    DB2023.findAll().then((DB2023) => {
        res.json(BazaDanych2023)
    })
  }); 
```
  
  kills node server with stack trace 
 ```
 FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF7969F012F node_api_throw_syntax_error+175407
 2: 00007FF796975AF6 SSL_get_quiet_shutdown+63110
 3: 00007FF796976EB2 SSL_get_quiet_shutdown+68162
 4: 00007FF79740FF24 v8::Isolate::ReportExternalAllocationLimitReached+116
 5: 00007FF7973FB282 v8::Isolate::Exit+674
 6: 00007FF79727D0DC v8::internal::EmbedderStackStateScope::ExplicitScopeForTesting+124
 7: 00007FF79727A2FB v8::internal::Heap::CollectGarbage+3963
 8: 00007FF797290533 v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath+2099
 9: 00007FF797290DDD v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath+93
10: 00007FF79729972A v8::internal::Factory::AllocateRaw+810
11: 00007FF7972B004A v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller+90
12: 00007FF7972B0333 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithMap+35
13: 00007FF797077936 v8::internal::HashTable<v8::internal::NameDictionary,v8::internal::NameDictionaryShape>::EnsureCapacity<v8::internal::Isolate>+246
14: 00007FF79707E92D v8::internal::BaseNameDictionary<v8::internal::NameDictionary,v8::internal::NameDictionaryShape>::Add+109
15: 00007FF796F87E46 v8::internal::Runtime::GetObjectProperty+982
16: 00007FF7974AD5C1 v8::internal::SetupIsolateDelegate::SetupHeap+558193
17: 00007FF717765E28
[nodemon] app crashed - waiting for file changes before starting...
```

```
--- Last few GCs --->

[15220:000002BB84AA1560]   181826 ms: Scavenge 4049.0 (4127.1) -> 4046.5 (4128.6) MB, 9.6 / 0.0 ms  (average mu = 0.838, current mu = 0.305) allocation failure;
[15220:000002BB84AA1560]   181841 ms: Scavenge 4050.4 (4128.6) -> 4048.5 (4132.3) MB, 10.2 / 0.0 ms  (average mu = 0.838, current mu = 0.305) allocation failure;
[15220:000002BB84AA1560]   183076 ms: Scavenge 4053.8 (4132.3) -> 4051.4 (4150.1) MB, 1229.4 / 0.0 ms  (average mu = 0.838, current mu = 0.305) allocation failure;
```

**Comment by ephys** [comment]
That does not indicate that there is a memory leak, only that loading this many rows at once and stringifying them as JSON requires too much memory. You need to either paginate or [stream](https://github.com/sequelize/sequelize/issues/10347) the data

**Comment by hesar** [comment]
sorry, for late response but you are completely right, browser was not able to handle such big chunk of data and was crashing
so advice for anyone willing to get huge data set at once, use other options like writing to file (e.g. you wanna export csv or similar)

**Comment by ozch** [comment]
having the same issue, server was crashing every-other day due to memory not being released, for now we just upped the resources and creating new connection on every incoming request and closing it afterwards,(its not pretty but then again its better than server crashing)

**Comment by p3v9d5ui** [comment]
Was anyone able to determine if this is an actual memory leak? We're seeing the same result, but we're not sure if the problem is Sequelize itself or how we use it.

**Comment by hesar** [comment]
> Was anyone able to determine if this is an actual memory leak? We're seeing the same result, but we're not sure if the problem is Sequelize itself or how we use it.

it wasnt leak for me, it was MY mistake

**Comment by SpeedyCraftah** [comment]
> I think I can confirm.
> 
> ![image](https://user-images.githubusercontent.com/45142584/105069488-eab1de00-5a79-11eb-8d5e-52bcfdb71aa1.png) ![image](https://user-images.githubusercontent.com/45142584/105069492-ebe30b00-5a79-11eb-84e8-5b2a1462a9d5.png)

As for my issue, I think I didn't take into account JS's garbage collection nature, if I were to run `gc.gc()` between these benchmarks, the number would likely stay the same. JS only starts purging memory once it starts running low or it is a convenient time.

**Comment by ozch** [comment]
@p3v9d5ui 
by any chance are you using pools with min open connection.

Disabled pooling and try closing the connection after each read, you will see the difference.

we ended up using query interface to get sql queries from Sequelize and started running it through node pg 

> Was anyone able to determine if this is an actual memory leak? We're seeing the same result, but we're not sure if the problem is Sequelize itself or how we use it.

**Comment by p3v9d5ui** [comment]
Hi @ozch . Yes, we are using connection pools, but our min/max connections setting are quite low. In fact I reduced both to just a couple of connections, and the problem still occurs. Also, this is a problem which only started recently, and those pool settings have not changed. Yeah, I guess that not using pools could help, but to me that's mitigating a symptom without identifying/solving the underlying problem.

**Comment by SippieCup** [comment]
disabling pooling entirely cause our OOM crashes to stop. Seems like some of the connections are not releasing memory. What is very strange is the fact that there is still a lot of headroom on pod it runs in.

**Comment by sk-tjdownes** [comment]
We're also experiencing this in our apps, consistently. This is happening on 5 different services. We are going to work on providing a reproducible example, as replacing 2 years worth of Sequelize code in our 5 apps isn't feasible. It may take us a few weeks to come up with a repro, since we can't start this work until mid-Feb.

**Comment by ozch** [comment]
@sk-tjdownes @p3v9d5ui @SippieCup have "temporary" a work-around for you guys, hope it helps, use queryGenerator.selectQuery and then run the query through a separate sql client with pools (i.e pg) so all you reads will through there (works for postgres with pg-node not sure about mysql).

depending on you architecture, you might wanna either create a utility function memSafeFindAll(Model, {...findAllParams}) or patch the Sequelize model to add a new function in there.

**Comment by sk-tjdownes** [comment]
> have "temporary" a work-around for you guys

Thanks, we'll look into this soon. I appreciate the heads up
